### ISSUE TITLE
Broadcast vs Send/Receive Performance Difference

### ISSUE THREAD

**Issue by bhetherman-bbn** [NVIDIA/nccl#568]
Hey All,

I have a DGXA100 box where I was running some NCCL tests. I ran a both a send/receive and a broadcast with 2 GPUs. I would expect the two to have basically the same performance however they were pretty different. It depended on the message size. At times broadcast was significantly faster and at times it was slower than send/receive. Also the latency at small message sizes was significantly higher for send/receive (16.5 us) than for broadcast (6.15 us).

I attached a plot I made that shows the ratio of the broadcast time over the time it takes to send/receive a message between two GPUs at different message sizes to show the behavior Im talking about. 

This leads me to a couple questions:

- Why is the latency so much higher for small messages when doing send/receive?
- What is going on as the message size increases to cause the difference to have the pattern shown in the plot? Do the different methods use different numbers of threads or do they break the message up differently, something else?

Any insight would be helpful! Thanks!

![Ratio of Broadcast Over SendRecieve vs Message Size](https://user-images.githubusercontent.com/86794081/133728028-bbb423fb-a819-44c2-83fe-ebb9851ca595.png)
The black line at 1 represents the time send/receive took.

For reference I used these params for the nccl tests when collecting the data "-t 2 -g 1 -n 1000 -p 1 -b 2 -e 8589934592 -f 2 -c 0 -r 0"

cc: @jkhoury

**Comment by sjeaugey** [comment]
Send/Recv and Broadcast are implemented in a very different way, because Send/Recv is optimized for alltoall (one rank communicating with all others). So it's not surprising to see such a difference, although obviously we'd want to eventually have send/recv be better or equal to broadcast.

**Comment by bhetherman-bbn** [comment]
Thanks for the reply @sjeaugey. That is good to know that they are implemented differently but I still dont fully understand why that is. At some level isnt all the data flowing p2p aka send/receive? 

Is there anything specific you can think of about those implementations that would help explain the difference? Why do we need a different control flow for alltoall vs broadcast? Is it hardware limitations or something else?

**Comment by sjeaugey** [comment]
Yes both will use the same underlying communication techniques, but they might use a different number of SMs, have different launch overhead, and use different protocols (Send/Recv only use `Simple` while Broadcast can use all 3 `LL,LL128,Simple`). So it is expected to see better latency for broadcast and eventually converge to the same bandwidth.

What I can't explain is the peak in the middle. But benchmarking is a tricky subject, especially for broadcast (and even more for -n 1000) because it only flows in one direction so it is subject to buffering effects. It could also simply be due to bad tuning where we don't use the right protocol at the right size (e.g. we switch from LL to LL128 too late).

**Comment by bhetherman-bbn** [comment]
Thanks for the additional information. That is helpful. Do you know if broadcast generally uses more SMs than send/receive? Does it depend on the number of node in the broadcast or on which proto is used?

**Comment by sjeaugey** [comment]
Broadcast and send/recv could use different number of SMs. Intra-node, send/recv might use more SMs than broadcast. But inter-node, send/recv would use only one SM per pair of ranks, although an alltoall would use more SMs than broadcast for a large number of ranks.

This is all dependent on the topology, i.e. the number of NVLinks (intra-node) or NICs (inter-node).

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
